File: src/disassemble.zig
const std = @import("std");
const root = @import("root");

const Forth = @import("forty/forth.zig");

// ----------------------------------------------------------------------
// Forty interop
// ----------------------------------------------------------------------

pub fn defineModule(forth: *Forth) !void {
    try forth.defineNamespace(@This(), .{.{ "disassemble", "disasm", "disassemble instruction, replace TOS with next addr" }});
}

// ----------------------------------------------------------------------
// Implementation
// ----------------------------------------------------------------------
extern fn disassemble_stub(addr: u64) u64;

pub fn disassemble(addr: u64) u64 {
    return disassemble_stub(addr);
}

File: attic/disassemble.c
#include "disasm/cstub.h"
#include "printf.h"
#include "disasm/aarch64.h"

uint64_t disassemble_stub(uint64_t addr) {
  char buf[256];
  uint64_t next = disasm(addr, buf);
  printf_("%08x    %s\n", addr, buf);
  return next;
}
